aws/cloudformation/cloud_formation_stack.yml.erb

```diff
@@ -76,7 +76,82 @@ Parameters:
 Resources:
+  # Stack-specific IAM permissions applied to both daemon and frontends.
+  CDOPolicy:
+    Type: AWS::IAM::ManagedPolicy
+    Properties:
+      Description: !Sub "Application permissions for ${AWS::StackName}."
+      PolicyDocument:
+        Statement:
+          # Read-only access to bootstrap scripts.
+          - Effect: Allow
+            Action: 's3:GetObject'
+            Resource: 'arn:aws:s3:::cdo-dist/<%=environment%>/*'
+          # Instance-bootstrap CloudFormation hook.
+          - Effect: Allow
+            Action: 'cloudformation:SignalResource'
+            Resource: !Ref AWS::StackId
+          # Forward syslog to CloudWatch Logs via cdo-cloudwatch-logger.
+          - Effect: Allow
+            Action:
+              - 'logs:CreateLogGroup'
+              - 'logs:CreateLogStream'
+              - 'logs:PutRetentionPolicy'
+              - 'logs:PutLogEvents'
+            Resource:
+              - !Sub "arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:<%=environment%>-syslog"
+              - !Sub "arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:<%=environment%>-syslog:log-stream:*"
+              - !Sub "arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:${AWS::StackName}"
+              - !Sub "arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:${AWS::StackName}:log-stream:*"
+          # Put custom metrics to CloudWatch.
+          - Effect: Allow
+            Action: 'cloudwatch:PutMetricData'
+            Resource: '*'
+          # Frontend-bootstrap lifecycle hooks.
+          - Effect: Allow
+            Action: 'autoscaling:CompleteLifecycleAction'
+            Resource: !Sub "arn:aws:autoscaling:${AWS::Region}:${AWS::AccountId}:autoScalingGroup:*:autoScalingGroupName/Frontends-${AWS::StackName}"
+          # Read/Write DCDO and Gatekeeper tables.
+          # TODO: Import resources into stack.
+          - Effect: Allow
+            Action:
+              - 'dynamodb:GetItem'
+              - 'dynamodb:PutItem'
+              - 'dynamodb:Scan'
+            Resource:
+              - !Sub "arn:aws:dynamodb:${AWS::Region}:${AWS::AccountId}:table/<%=CDO.dcdo_table_name%>"
+              - !Sub "arn:aws:dynamodb:${AWS::Region}:${AWS::AccountId}:table/<%=CDO.gatekeeper_table_name%>"
+          # Write analysis events to Firehose.
+          # TODO: Import resources into stack.
+          - Effect: Allow
+            Action: 'firehose:PutRecord'
+            Resource: !Sub "arn:aws:firehose:${AWS::Region}:${AWS::AccountId}:deliverystream/analysis-events"
+          # General s3 access.
+          # TODO: Further restrict permissions to grant least privilege.
+          - Effect: Allow
+            Action: 's3:*'
+            Resource: '*'
 <% if frontends -%>
+  FrontendRole:
+    Type: AWS::IAM::Role
+    Properties:
+      AssumeRolePolicyDocument:
+        Statement:
+          - Effect: Allow
+            Action: 'sts:AssumeRole'
+            Principal: {Service: ec2.amazonaws.com}
```

Comment:
I might add an erb-macro for the standard `AssumeRolePolicyDocument` boilerplate here, something like `<%=service_role 'ec2' %>` could replace these five lines for most aws-service-role cases.

Reply:
done